bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,488 @@
+pub struct DataSource {
+    reader: Box<dyn Read>,
+
+    total_size: usize,
+    cache_start_point: usize,
+    // cache size may be smaller than cache.len()
+    cache_size: usize,
+    cache: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub struct Cursor {
+    pub offset: usize,
+    pub size: usize,
+    pub data_source: Rc<RefCell<DataSource>>,
+}
+
+pub struct Union {
+    pub item_id: usize,
+    pub cursor: Cursor,
+}
+
+pub fn read_at(cur: &Cursor, buf: &mut [u8]) -> Result<usize, Error> {
+    let read_len = min(cur.size, buf.len());
+    let ds = &mut *cur.data_source.borrow_mut();
+    if read_len > ds.cache.len() {
+        return ds.reader.read(buf, cur.offset);
+    }
+    if cur.offset < ds.cache_start_point
+        || (cur.offset + read_len) > (ds.cache_start_point + ds.cache_size)
+    {
+        let reader = &ds.reader;
+        let size = reader.read(&mut ds.cache[..], cur.offset).unwrap();
+        if size < read_len {
+            return Err(Error::Read(format!(
+                "read_at `if size({}) < read_len({})`",
+                size, read_len
+            )));
+        }
+        ds.cache_size = size;
+        ds.cache_start_point = cur.offset;
+
+        if ds.cache_size > ds.cache.len() {
+            return Err(Error::Read(format!(
+                "read_at `if ds.cache_size({}) > ds.cache.len()({})`",
+                ds.cache_size,
+                ds.cache.len()
+            )));
+        }
+    }
+    if cur.offset < ds.cache_start_point || (cur.offset - ds.cache_start_point) > ds.cache.len() {
+        return Err(Error::Read(
+            "read_at `if cur.offset < ds.start_point || ...`".into(),
+        ));
+    }
+    let read_point = cur.offset - ds.cache_start_point;
+    if read_point + read_len > ds.cache_size {
+        return Err(Error::Read(
+            "read_at `if read_point + read_len > ds.cache_size`".into(),
+        ));
+    }
+    buf.copy_from_slice(&ds.cache[read_point..(read_point + read_len)]);
+    Ok(read_len)
+}
+
+impl Cursor {
+    /**
+    cache_size: normally it can be set to MAX_CACHE_SIZE(2K)
+    total_size: the size of cursor. If it's set a smaller value,
+    `out of bound` will occur when `reader` try to read the data beyond that.
+    reader: interface to read underlying data
+     */
+    pub fn new(total_size: usize, reader: Box<dyn Read>) -> Self {
+        let data_source = DataSource {
+            reader,
+            total_size,
+            cache_start_point: 0,
+            cache_size: 0, // when created, cache is not filled
+            cache: vec![0u8; MAX_CACHE_SIZE],
+        };
+        Cursor {
+            offset: 0,
+            size: total_size,
+            data_source: Rc::new(RefCell::new(data_source)),
+        }
+    }
+
+    pub fn add_offset(&mut self, offset: usize) {
+        self.offset = self.offset.checked_add(offset).unwrap();
+    }
+
+    pub fn sub_size(&mut self, shrink_size: usize) {
+        self.size = self.size.checked_sub(shrink_size).unwrap();
```

Comment:
same issue as `add_offset`, suggest to add unit test to cover these cases.

Reply:
`unwrap` is very difficult to debug when things go wrong, especially during unit testing. Building test cases to cover this error is easy, for example, define a table with two fields:
```
table Table1 {
    f1: Bytes,
    f2: Bytes,
}
```

and construct an invalid slice:
```
    let data = vec![
        0x17, 0x00, 0x00, 0x00,
        0x10, 0x00, 0x00, 0x00, 0x0c, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00,
        0x03, 0x00, 0x00, 0x00, 0xab, 0xcd, 0xef,
    ];
    let cursor = Cursor::new(data.len(), Box::new(data));
    let t = Table1 { cursor };
```

you will get an unwrap on None error then call `t.f1()`.

Reply:
`unwrap` is very difficult to debug when things go wrong, especially during unit testing. Building test cases is easy, for example, define a table with two fields:
```
table Table1 {
    f1: Bytes,
    f2: Bytes,
}
```

and init with an invalid slice:
```
    let data = vec![
        0x17, 0x00, 0x00, 0x00,
        0x10, 0x00, 0x00, 0x00, 0x0c, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00,
        0x03, 0x00, 0x00, 0x00, 0xab, 0xcd, 0xef,
    ];
    let cursor = Cursor::new(data.len(), Box::new(data));
    let t = types_api2::LazyReaderTable1 { cursor };
```

you will get an unwrap error on `None` of checked_sub result when calling `t.f1()` and `t.f2()`, it's very difficult to debug for contract. User cannot use customized error code to match this error unless we returning meaningful error here.

Reply:
Added checks on sub_size.